drivers/serial/uart_renesas_ra8_sci_b.c

```diff
@@ -0,0 +1,1164 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_ra8_uart_sci_b
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+#include "r_sci_b_uart.h"
+#include "r_dtc.h"
+#include "r_transfer_api.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ra8_uart_sci);
+
+#if defined(CONFIG_UART_ASYNC_API)
+void sci_b_uart_rxi_isr(void);
+void sci_b_uart_txi_isr(void);
+void sci_b_uart_tei_isr(void);
+void sci_b_uart_eri_isr(void);
+#endif
+
+struct uart_ra_sci_b_config {
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct uart_ra_sci_b_data {
+	const struct device *dev;
+	struct st_sci_b_uart_instance_ctrl sci;
+	struct uart_config uart_config;
+	struct st_uart_cfg fsp_config;
+	struct st_sci_b_uart_extended_cfg fsp_config_extend;
+	struct st_sci_b_baud_setting_t fsp_baud_setting;
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN)
+	uart_irq_callback_user_data_t user_cb;
+	void *user_cb_data;
+	uint32_t csr;
+#endif
+#if defined(CONFIG_UART_ASYNC_API)
+	/* RX */
+	struct st_transfer_instance rx_transfer;
+	struct st_dtc_instance_ctrl rx_transfer_ctrl;
+	struct st_transfer_info rx_transfer_info;
+	struct st_transfer_cfg rx_transfer_cfg;
+	struct st_dtc_extended_cfg rx_transfer_cfg_extend;
+	struct k_work_delayable rx_timeout_work;
+	size_t rx_timeout;
+	uint8_t *rx_buffer;
+	size_t rx_buffer_len;
+	size_t rx_buffer_cap;
+	size_t rx_buffer_offset;
+	uint8_t *rx_next_buffer;
+	size_t rx_next_buffer_cap;
+
+	/* TX */
+	struct st_transfer_instance tx_transfer;
+	struct st_dtc_instance_ctrl tx_transfer_ctrl;
+	struct st_transfer_info tx_transfer_info;
+	struct st_transfer_cfg tx_transfer_cfg;
+	struct st_dtc_extended_cfg tx_transfer_cfg_extend;
+	struct k_work_delayable tx_timeout_work;
+	size_t tx_timeout;
+	uint8_t *tx_buffer;
+	size_t tx_buffer_len;
+	size_t tx_buffer_cap;
+
+	uart_callback_t async_user_cb;
+	void *async_user_cb_data;
+#endif
+};
+
+static int uart_ra_sci_b_poll_in(const struct device *dev, unsigned char *c)
+{
+	struct uart_ra_sci_b_data *data = dev->data;
+	R_SCI_B0_Type *reg = data->sci.p_reg;
+
+	/* Check if async reception was enabled */
+	if (IS_ENABLED(CONFIG_UART_ASYNC_API) && reg->CCR0_b.RIE) {
+		return -EBUSY;
+	}
+
+	if (IS_ENABLED(CONFIG_UART_RA_SCI_B_UART_FIFO_ENABLE) ? reg->FRSR_b.DR == 0U
+							      : reg->CSR_b.RDRF == 0U) {
+		/* There are no characters available to read. */
+		return -1;
+	}
+
+	/* got a character */
+	*c = (unsigned char)reg->RDR;
+
+	return 0;
+}
+
+static void uart_ra_sci_b_poll_out(const struct device *dev, unsigned char c)
```

Comment:
One confirmation.

Is the `samples/drivers/uart/echo_bot/` sample work correctly?
This sample uses polling for sending and uses isr for receiving.

Reply:
@soburi : Yes, we confirmed samples/drivers/uart/echo_bot/ sample work correctly.

Reply:
I got it. (I got stuck when I tried it with RA4M1, so I checked it.)